src/Microsoft.VisualStudio.Threading/JoinableTask.cs

```diff
@@ -1105,8 +1128,29 @@ private bool TryDequeueSelfOrDependencies(bool onMainThread, ref HashSet<IJoinab
                             visited.Clear();
                         }
 
-                        if (TryDequeueSelfOrDependencies(this, onMainThread, visited, out work))
+                        var foundWork = TryDequeueSelfOrDependencies(this, onMainThread, visited, out work);
+
+                        HashSet<IJoinableTaskDependent>? visitedNodes = visited;
+                        if (visitedNodes != null && this.PotentialUnreachableDependents != null)
                         {
+                            // We walked the dependencies tree and use this information to update the PotentialUnreachableDependents list.
+                            this.PotentialUnreachableDependents.RemoveWhere(n => visitedNodes.Contains(n));
```

Comment:
This looks like an `O(n*m)` algorithm right here. Can we use `this.PotentialUnreachableDependents.ExceptWith(visitedNodes)` instead as that might be implemented more optimally, and may avoid a closure and delegate allocations.

Reply:
I don't agree. The visitedNodes is also a hashSet, so it is O(n*log(m)) vs O(log(n)*m) because ExceptWith takes an enumerable, and enumerate through them to remove items in the current set. However, PotentialUnreachable is usually a far smaller set than visitedNodes (which includes all reachable nodes in the dependency graph) so I think RemoveWhere is actually cheaper for a large graph, although it does pay the overhead of a closure.